### ISSUE TITLE
EQL: Decide on pagination mechanism and field

### ISSUE THREAD

**Issue by costin** [elastic/elasticsearch#52913]
Ticket #49634 proposed using `search_after` parameter as a way to pass information between requests to continue retrieving results.
However when it comes to sequences and joins, this seems inadequate to handle the information needed to _resume_ a query plus it requires the initial query to be sent again (which is unfortunate).

An alternative that is worth exploring is using a scroll/cursor like approach in which the resuming information is sent back to the client so a subsequent requests only sends that information back, without any other parameters.

This issue is raised to acknowledge the identified gap in the initial proposal and gather further discussions once development touches this topic.

**Comment by costin** [comment]
Pinging @tsg again for awareness as users of the API above.

**Comment by jimczi** [comment]
> However when it comes to sequences and joins, this seems inadequate to handle the information needed to resume a query plus it requires the initial query to be sent again (which is unfortunate).

Can you elaborate on that ? I don't see why `search_after` is not adequate to resume a query. It's a nice way in my opinion to move the state of the search to the client so that you **can** resume a query. I understand that for sequences, more work is needed to handle the state of individual queries but I don't see why we shouldn't use `search_after` as the basis to provide sorted streams.

**Comment by costin** [comment]
`search_after` at Elasticsearch query level is great for pagination.
But a similar parameter at EQL level is likely to not contain all the information needed to resume a sequence or join page. Internally EQL would use `search_after` to continue individual queries but assembling them into a next page is likely to extra information such the join key type, value in each query.

**Comment by imotov** [comment]
On the high level pagination means an ability to continue executing the physical plan where we left off. The physical plan for EQL in general consists of a set of one or more DSL queries and a state machine that walks (and leapfrogs) the results of the queries generating the results that  can be then post processed. In other words for each consecutive execution we need:

- Elasticsearch DSL queries
- Elasticsearch queries state (search_after keys or scroll_id)
- State machine configuration
- State machine state (unless we can always assume initial state)
- Postprocessing configuration

While designing the solution we would like to address the following concerns:

- Simple API
- Robustness (how should we react if the user doesn’t use API correctly)
- Server-side state (we would like to eliminate the server side state if possible)
- Upgradability or at least robustness error handling in case the the state sent to the user is no longer compatible with the current version of elasticsearch

For a multi page responses we will need to send a state of execution to a client so the client can send us this state back. There are 2 main approaches here. We can either encode all necessary information as a mega state key and send it to a client (similar to scroll API or SQL API) or we can encode only a minimal amount of information and ask a client to resent a complete request again for the next page so we can recreate necessary state (similar to search_after API). 

In case of a mega state approach we would parse EQL once and then encode all 5 elements ( elasticsearch queries, search_after keys, state machine state and configuration, postprocessing configuration)  into a key. 

In the second approach with a minimal key, we would recreate elasticsearch queries, state machine configuration and postprocessing configuration from re-supplied EQL and we will store elasticearch search_after keys and the state machine state if needed as the state key. 

The biggest issue with the first approach is that the key might be much bigger since EQL tends to be much less verbose compared to the DSL that results from it. The biggest issue with minimal state sent to the client is to deal with possibly inconsistent state. The inconsistency can be caused by changes into EQL between invocations or changes to optimization logic of EQL that will result in different DSL queries due to hitting a different ES version. 

I have been thinking about both approaches for a while, and I think I prefer the second approach primarily because it will involve much less serialization code that will have to be written. The consistency issue could be addressed by fingerprinting the resulted queries and verifying that EQL is still parsed into the same queries that the client supplied state was generated for. 

To summarize, unless I hear otherwise, I am planning to recreate the physical plan for every page from supplied EQL, and only send search_after keys and potentially small executor state back to the client. The consistency will be verified by fingerprinting the DSL queries.

**Comment by costin** [comment]
I prefer the former approach, fully serialized form, and I think it has some distinct advantages over the latter.

1. Usability wise, sending the request again on each page plus some "internal state" is leaky as it's a mixture of public API which is user-defined and internal state, which is server-defined.

2. Recreating the plans of the same query is wasteful - the field resolution means retrieving the mapping every time to check its correctness before creating the actual search queries. Index templates and aliases can lead to large mappings which need to be processed over and over. 
In metrics and SIEM case, mappings with close to thousands of fields are common a pattern making that number even higher. For cross-cluster search, the effect is amplified.

3. A nice effect of doing the planning once, during the initial request is not having to worry about further inconsistencies during pagination. Even if the query is the same, the state of cluster can change which can lead to breaking pagination due to incorrect planning.
The semantics become either as well - once a query has been planned, the follow-up is just search.

4. Serialization wise, this would apply to the internal state anyway so the difference between the two approaches would be on the executing queries. As these are actual Elastic queryDSL, they are already serializeable(`Writable`) and manageable in size.
Are there any components in particular that you're concerned about?

**Comment by imotov** [comment]
> Are there any components in particular that you're concerned about?

FieldExtractions and Attributes are probably my main concerns here. At the moment I don't see clearly how to limit the serialization concern here so it doesn't spill from EQL into the QL land. But if we go this way, I am sure we can figure it out.

I agree with all your concerns here except the first one. I was actually hoping that we can make interface cleaner with an approach modelled after "search_after", it's basically the same request with a small addon, especially if we can avoid base64 serialization for this addon. With scroll_id approach, it is basically one very structured and easy to understand request followed by the same request that only contains one field with a huge base64 string in it. So, I think there is opportunity to make this interface "transparent" rather than "leaky". 

Ultimately, I can see benefits of both approaches, and can easily argue both sides. I am ok with going with full plan serialization route, but I want to make sure that we are all ok with consequences of that path, which is more serialization work and uglier (in my eyes, subject to a debate) API.

**Comment by costin** [comment]
> FieldExtractions and Attributes are probably my main concerns here

`Attribute`s should not be serialized - they are used only for the plan output which is then converted into `FieldExtractor`s, that is the structure that knows how to extract the needed information from the response.
In EQL since the `_source` is returned as is, the field extraction should be minimal and occur only for some fields (like join key) which should simplify things further. In SQL only `FieldExtractor`s are serialized and the query dsl - the plans are not since they are not needed once the query has been transpiled.

**Comment by imotov** [comment]
> Attributes should not be serialized

Would that mean that we will have half initialized plans on the second page, or your plan is to have a specialized second page plans that will only exist on the second and the following pages?

**Comment by costin** [comment]
The implementation is not yet done so nothing is certain however my expectation is that one the query plan is decided upon, the initial page and the rest will contain ES queries plus some metadata to compute the actual result.
There would be no need for the Logical or Physical plan and their dependencies in place - they served their purpose. The actual execution plan would consist of `ActionListener`s and `FieldExtractor`s (or something similar, essentially a manipulation of the response itself).

**Comment by imotov** [comment]
OK, just to make sure I understand the plan (no pun intended). So the `PhysicalPlan` will not implement `execute(...)` anymore. Instead it will produce some other "action execution plan" that will implement `execute(...)` and upon execution will produce `Results` for the first page which will contain a serializable state that can be later rehydrated into a new "action execution plan" for the second page, which can be executed again and so on?

**Comment by costin** [comment]
`PhysicalPlan`s `execute` is not called universally - in fact most plans are `Unplanned`.
That's because there are plans that can only be executed locally (in SQL in particular the `SYS` commands or `LocalRelation` meaning you can return the results without hitting ES) or plans that are only available in ES - these get removed as they are folded into a `EsQueryPlan` which internally builds the query and as execute, triggers the query to ES.

However none of the above matter for pagination in SQL as once the query plan gets executed it returns the page results plus a cursor that contains the next ES query to be executed plus the deserialization of the results.
Meaning SQL and the trees have done their job - from this point on it's just about running normal ES queries and parsing the response.

**Comment by imotov** [comment]
So what are you proposing?

**Comment by jpountz** [comment]
Is my assumption correct that this discussion is only about the case when we return data to the client, and they later ask for the next page? The case when we need to consume all matches of an EQL query in order to feed a pipe for instance is not in the scope of this discussion?

If the above assumption is correct, I like Igor's proposal of having a minimal key better, requiring the client to re-send the query, and recreating the state on the server side. This would be consistent with the `_search` API and its `search_after` parameter. I also believe it would be a bit simpler, which I care about because I don't think that pagination is going to be a very common use-case.

Complex state keys also worry me backward-compatibility-wise. For instance think of the case of a 8.x coordinating node handling the initial request, and then the client asking for the next page to a 7.x coordinating node. Changing the format of this key is going to be challenging, so I'd rather follow the approach that would require the simplest state key.

**Comment by costin** [comment]
@tsg could you please take a look at the discussion above?
We'd like to know whether, for the most part, SIEM would typically navigate across multiple pages or be interested mainly in the first set of results. Especially if returning the results might not be immediate (think 10-20s or more for complicated queries or large datasets).

**Comment by tsg** [comment]
I'm thinking of two potential use-cases:

* When you create a new rule (backed by an EQL search) you might want to execute this rule historically to see what matches you get. A sub-case here, you might want to also generate signals for the historical results. So you might want to run a complex EQL for the last 1 month or so potentially returning a large number of hits (>1000).  I expect this rule will execute fully in the background. I'm not sure if async search and intermediate results have us covered there, or if we'd need to combine it with paging.

* For normal searches (Timeline or Discover when it will eventually support EQL) we'd want to allow pagination. My ideal UX for it would be automatic/infinite scrolling for the queries that are fast and an explicit "load more" button for queries that are slow + the option to put in background when needed. For the infinite scroll, it helps when we can have a reasonable guarantee that the results don't contain duplicates or gaps, otherwise the client code gets more complex.

Especially for the first use case, I'd be curious to learn how it mixes with async search. Do we need pagination at all in that case?

**Comment by imotov** [comment]
@tsg do you have any guesstimates on on how often we expect to paginate byong the first page of results across all use cases? Like is it 50% overall cases? 10%? 1%? 

> Especially for the first use case, I'd be curious to learn how it mixes with async search. Do we need pagination at all in that case?

I don't think we have finalized design for that yet. Do you have any particular preferences? What are you planning to do with 1000s of results in the first scenario?

**Comment by tsg** [comment]
> I don't think we have finalized design for that yet. Do you have any particular preferences? What are you planning to do with 1000s of results in the first scenario?

We would be re-indexing them as "signals", maybe with some filtering / enhancement. I _think_ it would be nice if the async API could send partial results, so we alleviate the need for paging in that case. But I don't know if that adds a lot of complexity on the ES side.

>@tsg do you have any guesstimates on on how often we expect to paginate byong the first page of results across all use cases? Like is it 50% overall cases? 10%? 1%?

Across use cases, it's hard to say because it depends on the other question. On UX use cases, I'd expect pagination to be needed ~ 20% of time.

**Comment by imotov** [comment]
Thanks for the feedback. Based on our discussion it sounds like the pagination might not be as critical for the first release as we originally thought. So, we are going to deprioritized this issue for now in favour of to asynchronous execution.

**Comment by costin** [comment]
So far there hasn't been any request for pagination. Further more pagination is currently possible through the `filter` parameter. As such, I'm closing this ticket since it's clear this functionality is *not* required for GA.

**Comment by Sankgreall** [comment]
I'm surprised there hasn't been more discussion/requests around pagination for EQL. I'm in the middle of migrating from MongoDB to Elastic, but without pagination I'm not sure Elastic can support the requirement.

If I make a large threat hunting query with EQL that could contain thousands or even tens of thousands of results, I need to return that data in pages because I need this service to operate with a low memory profile.

You've suggested that pagination is possible through the `filter` parameter, but I am not sure this is correct. To achieve this, I would sort my data both on the `@timestamp' and '_id' fields, and then could pass the values from the previous request into the filter parameter for the next query.

But this doesn't work for several reasons, including:

- The EQL `query` paramater does not support the `_id` field, so I can't do something like this: 
```
        if(lastTimestamp && lastId) {
            body.query += ` and (@timestamp > "${lastTimestamp}" or (@timestamp == "${lastTimestamp}" and _id > "${lastIid}"))`;
        }
``` 
- The DSL `filter` parameter does not support using `range` with the `_id` field, so I can't do something like this:
```
        if(lastTimestamp && lastId) {
            eqlQuery.body.filter = {
                bool: {
                    should: [
                        {range: {"@timestamp": {gt: lastTimestamp}}},
                        {
                            bool: {
                                must: [
                                    {term: {"@timestamp": lastTimestamp}},
                                    // Please replace field_name with your sortable unique field
                                    {range: {"_id": {gt: lastId}}}
                                ]
                            }
                        }
                    ]
                }
            };
        }
```
- Although I could duplicate my `_id` field and add it to the document index, for either of the two options above to work my id values would have to be incremental. I am dealing with random UUIDs, which means these methods of pagination won't work if data is continually being added to the index as I am querying it.

Am I missing something, because this seems to be quite an important issue? How are others dealing with this, because almost all the EQL queries I intend to migrate will return significant numbers of documents - but I can't immediately see how I can support pagination here.
